File: 993 - Cousins in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int xdep, ydep, xval, yval;
    TreeNode* xparent, * yparent;

public:
    bool isCousins(TreeNode* root, int x, int y) {
        xdep = ydep = -1;
        xval = x;
        yval = y;
        getDep(root, 0, nullptr);
        return xdep == ydep && xparent != yparent;
    }
    void getDep(TreeNode* root, int curDep, TreeNode* parent) {
        if (root == nullptr) return;
        if (root->val == xval) {
            xdep = curDep;
            xparent = parent;
        }
        else if (root->val == yval) {
            ydep = curDep;
            yparent = parent;
        }
        else {
            getDep(root->left, curDep + 1, root);
            getDep(root->right, curDep + 1, root);
        }
    }
};

File: 993 - Cousins in Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
let xdep, ydep, xparent, yparent, vx, vy;
/**
 * @param {TreeNode} root
 * @param {number} x
 * @param {number} y
 * @return {boolean}
 */
var isCousins = function (root, x, y) {
    vx = x;
    xdep = -1;
    ydep = -1;
    vy = y;
    getDep(root, 0, null);
    return xdep == ydep && xparent != yparent;
};


function getDep(root, curDep, parent) {
    if (root == null) return;
    if (root.val == vx) {
        xdep = curDep;
        xparent = parent;
    } else if (root.val == vy) {
        ydep = curDep;
        yparent = parent;
    } else {
        getDep(root.left, curDep + 1, root);
        getDep(root.right, curDep + 1, root);
    }
}